File: mss32/src/textids.cpp
/*
 * This file is part of the modding toolset for Disciples 2.
 * (https://github.com/VladimirMakeev/D2ModdingToolset)
 * Copyright (C) 2021 Stanislav Egorov.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

#include "textids.h"
#include "log.h"
#include "scripts.h"
#include "utils.h"
#include <fmt/format.h>

namespace hooks {

void readOwnResourceTextIds(const sol::table& table,
                            TextIds::Events::Conditions::OwnResource& value)
{
    value.tooMany = table.get_or("tooMany", std::string());
    value.mutuallyExclusive = table.get_or("mutuallyExclusive", std::string());
}

void readGameModeTextIds(const sol::table& table, TextIds::Events::Conditions::GameMode& value)
{
    value.tooMany = table.get_or("tooMany", std::string());
    value.single = table.get_or("single", std::string());
    value.hotseat = table.get_or("hotseat", std::string());
    value.online = table.get_or("online", std::string());
}

void readPlayerTypeTextIds(const sol::table& table, TextIds::Events::Conditions::PlayerType& value)
{
    value.tooMany = table.get_or("tooMany", std::string());
    value.human = table.get_or("human", std::string());
    value.ai = table.get_or("ai", std::string());
}

void readVariableCmpTextIds(const sol::table& table,
                            TextIds::Events::Conditions::VariableCmp& value)
{
    value.equal = table.get_or("equal", std::string());
    value.notEqual = table.get_or("notEqual", std::string());
    value.greater = table.get_or("greater", std::string());
    value.greaterEqual = table.get_or("greaterEqual", std::string());
    value.less = table.get_or("less", std::string());
    value.lessEqual = table.get_or("lessEqual", std::string());
}

void readConditionsTextIds(const sol::table& table, TextIds::Events::Conditions& value)
{
    auto conditions = table.get<sol::optional<sol::table>>("conditions");
    if (!conditions.has_value())
        return;

    auto ownResource = conditions.value().get<sol::optional<sol::table>>("ownResource");
    if (ownResource.has_value()) {
        readOwnResourceTextIds(ownResource.value(), value.ownResource);
    }

    auto gameMode = conditions.value().get<sol::optional<sol::table>>("gameMode");
    if (gameMode.has_value()) {
        readGameModeTextIds(gameMode.value(), value.gameMode);
    }

    auto playerType = conditions.value().get<sol::optional<sol::table>>("playerType");
    if (playerType.has_value()) {
        readPlayerTypeTextIds(playerType.value(), value.playerType);
    }

    auto variableCmp = conditions.value().get<sol::optional<sol::table>>("variableCmp");
    if (variableCmp.has_value()) {
        readVariableCmpTextIds(variableCmp.value(), value.variableCmp);
    }
}

void readEventsTextIds(const sol::table& table, TextIds::Events& value)
{
    auto events = table.get<sol::optional<sol::table>>("events");
    if (!events.has_value())
        return;

    readConditionsTextIds(events.value(), value.conditions);
}

void readLobbyTextIds(const sol::table& table, TextIds::Lobby& value)
{
    auto lobbyTable = table.get<sol::optional<sol::table>>("lobby");
    if (!lobbyTable.has_value())
        return;

    auto& lobby = lobbyTable.value();
    value.serverName = lobby.get_or("serverName", std::string());
    value.serverNotResponding = lobby.get_or("serverNotResponding", std::string());
    value.connectAttemptFailed = lobby.get_or("connectAttemptFailed", std::string());
    value.serverIsFull = lobby.get_or("serverIsFull", std::string());
    value.computeHashFailed = lobby.get_or("computeHashFailed", std::string());
    value.requestHashCheckFailed = lobby.get_or("requestHashCheckFailed", std::string());
    value.wrongHash = lobby.get_or("wrongHash", std::string());
    value.wrongRoomPassword = lobby.get_or("wrongRoomPassword", std::string());
}

void readGeneratorTextIds(const sol::table& table, TextIds::ScenarioGenerator& value)
{
    auto rsgTable = table.get<sol::optional<sol::table>>("generator");
    if (!rsgTable.has_value())
        return;

    auto& rsg = rsgTable.value();
    value.description = rsg.get_or("description", std::string());
    value.wrongGameData = rsg.get_or("wrongGameData", std::string());
    value.generationError = rsg.get_or("generationError", std::string());
    value.limitExceeded = rsg.get_or("limitExceeded", std::string());
}

void readInterfTextIds(const sol::table& table, TextIds::Interf& value)
{
    auto interf = table.get<sol::optional<sol::table>>("interf");
    if (!interf.has_value())
        return;

    value.sellAllValuables = interf.value().get_or("sellAllValuables", std::string());
    value.sellAllItems = interf.value().get_or("sellAllItems", std::string());
    value.critHitAttack = interf.value().get_or("critHitAttack", std::string());
    value.critHitDamage = interf.value().get_or("critHitDamage", std::string());
    value.ratedDamage = interf.value().get_or("ratedDamage", std::string());
    value.ratedDamageEqual = interf.value().get_or("ratedDamageEqual", std::string());
    value.ratedDamageSeparator = interf.value().get_or("ratedDamageSeparator", std::string());
    value.splitDamage = interf.value().get_or("splitDamage", std::string());
    value.modifiedValue = interf.value().get_or("modifiedValue", std::string());
    value.modifiedNumber = interf.value().get_or("modifiedNumber", std::string());
    value.modifiedNumberTotal = interf.value().get_or("modifiedNumberTotal", std::string());
    value.positiveBonusNumber = interf.value().get_or("positiveBonusNumber", std::string());
    value.negativeBonusNumber = interf.value().get_or("negativeBonusNumber", std::string());
    value.modifiersCaption = interf.value().get_or("modifiersCaption", std::string());
    value.modifiersEmpty = interf.value().get_or("modifiersEmpty", std::string());
    value.modifierDescription = interf.value().get_or("modifierDescription", std::string());
    value.nativeModifierDescription = interf.value().get_or("nativeModifierDescription",
                                                            std::string());
    value.drainDescription = interf.value().get_or("drainDescription", std::string());
    value.drainEffect = interf.value().get_or("drainEffect", std::string());
    value.overflowAttack = interf.value().get_or("overflowAttack", std::string());
    value.overflowText = interf.value().get_or("overflowText", std::string());
    value.dynamicUpgradeLevel = interf.value().get_or("dynamicUpgradeLevel", std::string());
    value.dynamicUpgradeValues = interf.value().get_or("dynamicUpgradeValues", std::string());
    value.durationDescription = interf.value().get_or("durationDescription", std::string());
    value.durationText = interf.value().get_or("durationText", std::string());
    value.instantDurationText = interf.value().get_or("instantDurationText", std::string());
    value.randomDurationText = interf.value().get_or("randomDurationText", std::string());
    value.singleTurnDurationText = interf.value().get_or("singleTurnDurationText", std::string());
    value.wholeBattleDurationText = interf.value().get_or("wholeBattleDurationText", std::string());
    value.infiniteAttack = interf.value().get_or("infiniteAttack", std::string());
    value.infiniteText = interf.value().get_or("infiniteText", std::string());
    value.removedAttackWard = interf.value().get_or("removedAttackWard", std::string());
    value.extraUnitStats = interf.value().get_or("extraUnitStats", std::string());
}

void initialize(TextIds& value)
{
    const auto path{hooks::scriptsFolder() / "textids.lua"};
    try {
        const auto env{executeScriptFile(path)};
        if (!env)
            return;

        const sol::table& table = (*env)["textids"];
        readInterfTextIds(table, value.interf);
        readEventsTextIds(table, value.events);
        readLobbyTextIds(table, value.lobby);
        readGeneratorTextIds(table, value.rsg);
    } catch (const std::exception& e) {
        showErrorMessageBox(fmt::format("Failed to read script '{:s}'.\n"
                                        "Reason: '{:s}'",
                                        path.string(), e.what()));
    }
}

const TextIds& textIds()
{
    static TextIds value;
    static bool initialized = false;

    if (!initialized) {
        initialize(value);
        initialized = true;
    }

    return value;
}

} // namespace hooks

File: Scripts/textids.lua
--[[
Text ids for Disciples 2 Rise of the Elves v3.01 mss32 proxy dll

Refer to the documentation to find exact meaning of each text id
https://github.com/VladimirMakeev/D2ModdingToolset/blob/master/README.md
--]]

textids = {
	-- Interface text ids from either TApp.dbf or TAppEdit.dbf
	interf = {
		-- Defines text id to use as "sell all valuables" message.
		-- The text must contain keyword "%PRICE%".
		-- Fallback text is "Do you want to sell all valuables? Revenue will be:\n%PRICE%"
		sellAllValuables = "X015TA0001",

		-- Defines text id to use as "sell all items" message.
		-- The text must contain keyword "%PRICE%".
		-- Fallback text is "Do you want to sell all items? Revenue will be:\n%PRICE%"
		sellAllItems = "",

		-- Defines format id for effect duration description in unit encyclopedia.
		-- The text must contain keywords "%DURATION%" and "%DURATIONVALUE%".
		-- Fallback text is "\fMedBold;%DURATION%:\t\fNormal;%DURATIONVALUE%\n"
		durationDescription = "",

		-- Defines text id of effect duration in unit encyclopedia.
		-- Fallback text is "Duration".
		durationText = "",

		-- Defines text id of instant effect duration in unit encyclopedia.
		-- Fallback text is "Instant".
		instantDurationText = "",

		-- Defines text id of random effect duration in unit encyclopedia.
		-- Fallback text is "Random".
		randomDurationText = "",

		-- Defines text id of single-turn effect duration in unit encyclopedia.
		-- Fallback text is "Single turn".
		singleTurnDurationText = "",

		-- Defines text id of whole-battle effect duration in unit encyclopedia.
		-- Fallback text is "Whole battle".
		wholeBattleDurationText = "",

		-- Defines text id to mark Critical Hit attacks in unit encyclopedia.
		-- Fallback text id is the standard "X160TA0017".
		critHitAttack = "",

		-- Defines text id to format Critical Hit text in unit encyclopedia.
		-- Despite its name, it is used to format both crit hit damage and power (chance to hit).
		-- The text must contain keywords "%DMG%" and "%CRIT%".
		-- Fallback text is "%DMG% (%CRIT%)".
		critHitDamage = "",

		-- Defines text id to format rated damage in unit encyclopedia.
		-- The text must contain keywords "%DMG%" and "%RATED%".
		-- Fallback text is "%DMG%, %RATED%".
		ratedDamage = "",

		-- Defines text id to format equal rated damage in unit encyclopedia.
		-- The text must contain keywords "%DMG%", %TARGETS% and "%RATED%".
		-- Fallback text is "%DMG%, (%TARGETS%x) %RATED%".
		ratedDamageEqual = "",

		-- Defines text id for rated damage separator in unit encyclopedia.
		-- Fallback text is ", ".
		ratedDamageSeparator = "",

		-- Defines text id to format split damage in unit encyclopedia.
		-- The text must contain keyword "%DMG%".
		-- Fallback text is "%DMG%, split between targets".
		splitDamage = "",

		-- Defines modified text representation in unit encyclopedia.
		-- Modified text includes attack name, source, reach, etc.
		-- The text must contain keyword "%VALUE%".
		-- Fallback text is "\c025;090;000;%VALUE%\c000;000;000;".
		modifiedValue = "",

		-- Defines modified number representation in unit encyclopedia.
		-- The text must contain keywords "%NUMBER%" and "%BONUS%".
		-- Fallback text is "%NUMBER% %BONUS%".
		modifiedNumber = "",

		-- Defines modified number with total representation in unit encyclopedia.
		-- Used for HP and XP bonus display.
		-- The text must contain keywords "%TOTAL%" and "%BONUS%".
		-- Fallback text is "%TOTAL% (%BONUS%)".
		modifiedNumberTotal = "",

		-- Defines positive bonus number representation in unit encyclopedia.
		-- The text must contain keywords "%NUMBER%" and "%SIGN%".
		-- Note that sign can be either + or - depending on stat.
		-- Fallback text is "\c025;090;000;%SIGN% %NUMBER%\c000;000;000;".
		positiveBonusNumber = "",

		-- Defines negative bonus number representation in unit encyclopedia.
		-- The text must contain keywords "%NUMBER%" and "%SIGN%".
		-- Note that sign can be either + or - depending on stat.
		-- Fallback text is "\c100;000;000;%SIGN% %NUMBER%\c000;000;000;".
		negativeBonusNumber = "",

		-- Defines modifiers list caption in unit encyclopedia.
		-- Fallback text is "\fMedBold;Effects:\fNormal;"
		modifiersCaption = "",

		-- Defines empty modifiers list content in unit encyclopedia.
		-- Fallback text id is "X005TA0676" (empty leader abilities).
		modifiersEmpty = "",

		-- Defines format id for modifier description in unit encyclopedia.
		-- The text must contain keyword "%DESC%".
		-- Fallback text is "\vC;%DESC%"
		modifierDescription = "",

		-- Defines format id for native modifier description in unit encyclopedia.
		-- The text must contain keyword "%DESC%".
		-- Fallback text is "\vC;\fMedBold;%DESC%\fNormal;"
		nativeModifierDescription = "",

		-- Defines format id for drain description in unit encyclopedia.
		-- The text must contain keywords "%DRAINEFFECT%" and "%DRAIN%".
		-- Fallback text is "\fMedBold;%DRAINEFFECT%:\t\fNormal;%DRAIN%\n"
		drainDescription = "",

		-- Defines text id of drain effect in unit encyclopedia.
		-- Fallback text id is the standard "X005TA0792".
		drainEffect = "",

		-- Defines text id to mark Drain Overflow attacks in unit encyclopedia.
		-- Fallback text is "Overflow".
		overflowAttack = "",

		-- Defines text id to format Drain Overflow text in unit encyclopedia.
		-- The text must contain keywords "%ATTACK%" and "%OVERFLOW%".
		-- Fallback text is "%ATTACK% (%OVERFLOW%)".
		overflowText = "",

		-- Defines text id to format dynamic upgrade level in unit encyclopedia.
		-- The text must contain keyword %STAT%.
		-- "%UPGLV%" is optional and can be ommited if you don't want to display upgrade level.
		-- Fallback text is "%STAT% (level-ups weaken at %UPGLV%)".
		dynamicUpgradeLevel = "",

		-- Defines text id to format dynamic upgrades text in unit encyclopedia.
		-- The text must contain keywords "%STAT%", %UPG1% and %UPG2%.
		-- Fallback text is "%STAT% (%UPG1% | %UPG2% per level)".
		dynamicUpgradeValues = "",
		
		-- Defines text id to format broken (removed) wards in unit encyclopedia.
		-- The text must contain keyword "%WARD%", or it can be an empty string to not display broken wards at all.
		-- Fallback text is "\fMedBold;\c100;000;000;%WARD%\c000;000;000;\fNormal;".
		removedAttackWard = "",

		-- Defines text id to mark Infinite attacks in unit encyclopedia.
		-- Fallback text is "Lasting".
		infiniteAttack = "",

		-- Defines text id to format Infinite text in unit encyclopedia.
		-- The text must contain keywords "%ATTACK%" and "%INFINITE%".
		-- Fallback text is "%ATTACK% (%INFINITE%)".
		infiniteText = "",
	},

	-- Text ids related to events logic
	events = {
		-- Text ids related to event conditions
		conditions = {
			-- Text ids for L_OWN_RESOURCE event condition
			ownResource = {
				-- Message to show when there are more than two conditions of type
				-- Fallback text is "At most two conditions of type "Own resource" is allowed per event."
				tooMany = "",
				-- Message to show when two conditions are mutually exclusive
				-- Fallback text is "Conditions of type "Own resource" are mutually exclusive."
				mutuallyExclusive = "",
			},
	
			-- Text ids for L_GAME_MODE event condition
			gameMode = {
				-- Message to show when there are more than one condition of type
				-- Fallback text "Only one condition of type "Game mode" is allowed per event."
				tooMany = "",
				-- Text id for single player game mode
				-- Fallback text "single player"
				single = "",
				-- Text id for hotseat game mode
				-- Fallback text "hotseat"
				hotseat = "",
				-- Text id for online game mode
				-- Fallback text "online"
				online = "",
			},
	
			-- Text ids for L_PLAYER_TYPE event condition
			playerType = {
				-- Message to show when there are more than one condition of type
				-- Fallback text "Only one condition of type "Player type" is allowed per event."
				tooMany = "",
				-- Text id for human player type
				-- Fallback text "human"
				human = "",
				-- Text id for AI player type
				-- Fallback text "AI"
				ai = "",
			},

			-- Text ids for L_VARIABLE_CMP event condition
			variableCmp = {
				-- Text id for equality comparison
				-- Fallback text "is equal to"
				equal = "",
				-- Text id for difference comparison
				-- Fallback text "is not equal to"
				notEqual = "",
				-- Fallback text "is greater than"
				greater = "",
				-- Fallback text "is greater or equal to"
				greaterEqual = "",
				-- Fallback text "is less than"
				less = "",
				-- Fallback text "is less or equal to"
				lessEqual = "",
			},
		},
	},

	-- Text ids related to custom lobby
	lobby = {
		-- Server name shown in network protocols list
		-- Fallback text "Lobby server"
		serverName = "",
		-- Shown when client could not connect due to server being unresponsive
		-- Fallback text "Failed to connect.\nLobby server not responding"
		serverNotResponding = "",
		-- Client's connection attemt failed
		-- Fallback text "Connection attempt failed"
		connectAttemptFailed = "",
		-- Client could not connect due to server being full
		-- Fallback text "Lobby server is full"
		serverIsFull = "",
		-- Game files hash failed to compute
		-- Fallback text "Could not compute hash"
		computeHashFailed = "",
		-- Client could not request hash check from server
		-- Fallback text "Could not request game integrity check"
		requestHashCheckFailed = "",
		-- Client hash is wrong
		-- Fallback text "Game integrity check failed"
		wrongHash = "",
		-- Player entered wrong room password
		-- Fallback text "Wrong room password"
		wrongRoomPassword = "",
	},
    
    generator = {
        -- Description text for randomly generated scenarios
        -- Fallback text "Random scenario based on template '%TMPL%'. Seed: %SEED%. Starting gold: %GOLD%. Roads: %ROADS%%. Forest: %FOREST%%."
        description = "",
        -- Generator could not process game data from dbf tables or .ff files
        -- Error details are logged in mssProxyError.log
        -- Fallback text "Could not read game data needed for scenario generator.\nSee mssProxyError.log for details"
        wrongGameData = "",
        -- Error occured during scenario generation
        -- Fallback text "Error during random scenario map generation.\nSee mssProxyError.log for details".
        generationError = "",
        -- Generator failed to create scenario after specified number of attempts
        -- Fallback text "Could not generate scenario map after %NUM% attempts.\nPlease, adjust template contents or settings"
        limitExceeded = "",
    },
}
